File: app/src/main/java/com/fr1014/mycoludmusic/ui/home/toplist/PlayListDetailAdapter.java
package com.fr1014.mycoludmusic.ui.home.toplist;

import android.text.TextUtils;
import android.view.View;
import android.widget.LinearLayout;

import androidx.constraintlayout.widget.ConstraintLayout;

import com.fr1014.frecyclerviewadapter.BaseAdapter;
import com.fr1014.frecyclerviewadapter.BaseViewHolder;
import com.fr1014.mycoludmusic.R;
import com.fr1014.mycoludmusic.musicmanager.Music;
import com.fr1014.mycoludmusic.utils.ScreenUtil;

/**
 * 创建时间:2020/9/5
 * 作者:fr
 * 邮箱:1546352238@qq.com
 */
public class PlayListDetailAdapter extends BaseAdapter<Music, BaseViewHolder> {
    //默认显示序号
    private boolean isDisplaySeq = true;

    //是否显示底部的View
    private boolean isDisplayMarginView = false;

    /**
     * @param isDisplaySeq 是否显示序号
     */
    public PlayListDetailAdapter(boolean isDisplaySeq) {
        this();
        this.isDisplaySeq = isDisplaySeq;
    }

    public PlayListDetailAdapter() {
        super(R.layout.item_playlist_detail, null);
    }

    private static final String TAG = "PlayListDetailAdapter";

    @Override
    protected void convert(BaseViewHolder holder, Music data) {
        if (isDisplaySeq) {
            holder.getView(R.id.cl_status).setVisibility(View.VISIBLE);
            holder.setText(R.id.tv_rank, holder.getLayoutPosition() + "");
        } else {
            holder.getView(R.id.cl_status).setVisibility(View.GONE);
        }
        holder.setText(R.id.tv_song_name, data.getTitle());

        //歌手 - 专辑
        if (TextUtils.isEmpty(data.getAlbum())) {
            holder.setText(R.id.tv_author, data.getArtist());
        } else {
            holder.setText(R.id.tv_author, data.getArtist() + " - " + data.getAlbum());
        }

        //是否为原唱
        if (TextUtils.equals(data.getOriginal(), "1")) {
            holder.getView(R.id.tv_prompt).setVisibility(View.VISIBLE);
            holder.setText(R.id.tv_prompt, mContext.getString(R.string.original));
            setAuthorLeftMargin(holder, 4);
        } else {
            holder.getView(R.id.tv_prompt).setVisibility(View.GONE);
            setAuthorLeftMargin(holder, 0);
        }

        //别名/介绍
        if (TextUtils.isEmpty(data.getSubTitle())) {
            holder.getView(R.id.tv_subtitle).setVisibility(View.GONE);
        } else {
            holder.getView(R.id.tv_subtitle).setVisibility(View.VISIBLE);
            holder.setText(R.id.tv_subtitle, data.getSubTitle());
        }

        //是否显示最底部的MarginView
        if (isDisplayMarginView) {
            holder.getView(R.id.margin_barsize).setVisibility(isShowDivider(holder) ? View.VISIBLE : View.GONE);
        }

    }

    private boolean isShowDivider(BaseViewHolder holder) {
//        Log.d(TAG, "----isShowDivider1: " + (holder.getLayoutPosition() == getDatas().size() - 1));
//        Log.d(TAG, "----isShowDivider2: " + "holder.getLayoutPosition():" + holder.getLayoutPosition());
//        Log.d(TAG, "----isShowDivider3: " + "getDatas().size() - 1:" + (getDatas().size() - 1));
        return getRealPosition(holder) == getDatas().size() - 1;
    }

    public void setDisplayMarginView(boolean displayMarginView) {
        isDisplayMarginView = displayMarginView;
    }

    private void setAuthorLeftMargin(BaseViewHolder holder, int dp) {
        LinearLayout.LayoutParams layoutParams = (LinearLayout.LayoutParams) holder.getView(R.id.tv_author).getLayoutParams();
        layoutParams.leftMargin = ScreenUtil.dp2px(mContext, dp);
        holder.getView(R.id.tv_author).setLayoutParams(layoutParams);
    }
}

File: app/src/main/java/com/fr1014/mycoludmusic/ui/search/paging2/PlayListDetailAdapter.kt
package com.fr1014.mycoludmusic.ui.search.paging2

import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.fr1014.frecyclerviewadapter.BaseViewHolder
import com.fr1014.mycoludmusic.R
import com.fr1014.mycoludmusic.SourceHolder
import com.fr1014.mycoludmusic.musicmanager.Music
import com.fr1014.mycoludmusic.ui.search.SearchViewModel
import com.fr1014.mycoludmusic.utils.ScreenUtil

/**
 * Create by fanrui on 2021/1/10
 * Describe:
 */
class PlayListDetailAdapter(private val mViewModel: SearchViewModel) : PagedListAdapter<Music, RecyclerView.ViewHolder>(DIFF_CALLBACK) {

    //是否显示底部的View
    private var isDisplayMarginView = false

    companion object{
        val DIFF_CALLBACK = object : DiffUtil.ItemCallback<Music>() {
            override fun areItemsTheSame(oldItem: Music, newItem: Music): Boolean {
                return oldItem == newItem
            }

            override fun areContentsTheSame(oldItem: Music, newItem: Music): Boolean {
                return (oldItem.title + oldItem.artist) == (newItem.title + newItem.artist)
            }
        }
    }


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
//        return when (viewType) {
//            R.layout.item_playlist_detail -> PlayListViewHolder.newInstance(parent).also { holder ->
//                holder.itemView.setOnClickListener {
//
//                }
//            }
//
//            else -> FooterViewHolder.newInstance(parent).also {
//                it.itemView.setOnClickListener {
//                    galleryViewModel.retry()
//                }
//        }
        return PlayListViewHolder.newInstance(parent).also { holder ->
            holder.itemView.setOnClickListener {
                when (SourceHolder.get().source) {
                    "酷我" -> mViewModel.getKWSongUrl(getItem(holder.adapterPosition) as Music)
                    "网易" -> mViewModel.getWYYSongUrl(getItem(holder.adapterPosition) as Music)
                }
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (holder.itemViewType) {
//            R.layout.gallery_footer -> (holder as FooterViewHolder).bindWithNetworkStatus(
//                    networkStatus
//            )
            else -> {
                val music = getItem(position) ?: return
                (holder as PlayListViewHolder).bindWithPhotoItem(music)
            }
        }
    }

    class PlayListViewHolder(itemView: View) : BaseViewHolder(itemView) {

        companion object {
            fun newInstance(parent: ViewGroup): PlayListViewHolder {
                val view = LayoutInflater.from(parent.context).inflate(R.layout.item_playlist_detail, parent, false)
                return PlayListViewHolder(view)
            }
        }

        fun bindWithPhotoItem(music: Music) {
            getView<ConstraintLayout>(R.id.cl_status).visibility = View.GONE
            setText(R.id.tv_song_name, music.title)
            //歌手 - 专辑
            if (TextUtils.isEmpty(music.album)) {
                setText(R.id.tv_author, music.artist)
            } else {
                setText(R.id.tv_author, music.artist + " - " + music.album)
            }

            //是否为原唱
            if (TextUtils.equals(music.original, "1")) {
                getView<View>(R.id.tv_prompt).visibility = View.VISIBLE
                setText(R.id.tv_prompt, itemView.context.getString(R.string.original))
                setAuthorLeftMargin(4)
            } else {
                getView<View>(R.id.tv_prompt).visibility = View.GONE
                setAuthorLeftMargin(0)
            }

            //别名/介绍
            if (TextUtils.isEmpty(music.subTitle)) {
                getView<View>(R.id.tv_subtitle).visibility = View.GONE
            } else {
                getView<View>(R.id.tv_subtitle).visibility = View.VISIBLE
                setText(R.id.tv_subtitle, music.subTitle)
            }

//            //是否显示最底部的MarginView
//            if (isDisplayMarginView) {
//                holder.getView<View>(R.id.margin_barsize).setVisibility(if (isShowDivider(holder)) View.VISIBLE else View.GONE)
//            }
        }

        private fun setAuthorLeftMargin(dp: Int) {
            val layoutParams = getView<View>(R.id.tv_author).layoutParams as LinearLayout.LayoutParams
            layoutParams.leftMargin = ScreenUtil.dp2px(itemView.context, dp.toFloat())
            getView<View>(R.id.tv_author).layoutParams = layoutParams
        }
    }

    fun setDisplayMarginView(displayMarginView: Boolean) {
        isDisplayMarginView = displayMarginView
    }
}